### PR TITLE
MDEV-15547 - auth_pam: add PAM_RHOST to pam info

### PR DESCRIPTION
**Pull request by grooverdan**
plugins.pam                              [ pass ]  46436

I submit this under the MCA.

### CODE REVIEW

**Comment by grooverdan** [comment]
<pre>
$ mysql-test/mtr --mysqld=--loose-pam-debug=1  --mem pam
Logging: /home/dan/repos/mariadb-server-10.3/mysql-test/mysql-test-run.pl  --mysqld=--loose-pam-debug=1 --mem pam
vardir: /home/dan/repos/build-mariadb-server-10.3-debug/mysql-test/var
Checking leftover processes...
Removing old var directory...
Creating var directory '/home/dan/repos/build-mariadb-server-10.3-debug/mysql-test/var'...
 - symlinking 'var' to '/dev/shm/var_auto_Wqhf'
Checking supported features...
MariaDB Version 10.3.5-MariaDB-debug
 - SSL connections supported
 - binaries are debug compiled
Collecting tests...
Installing system database...

==============================================================================

TEST                                      RESULT   TIME (ms) or COMMENT
--------------------------------------------------------------------------

worker[1] Using MTR_BUILD_THREAD 300, with reserved ports 16000..16019
plugins.pam                              [ pass ]     38
--------------------------------------------------------------------------
The servers were restarted 0 times
Spent 0.038 of 4 seconds executing testcases

Completed: All 1 tests were successful.


$ grep PAM_RHOST  mysql-test/var/log/mysqld.1.err            
PAM: pam_set_item(PAM_RHOST, 127.0.0.1)

$ mysql-test/mtr --mysqld=--loose-pam-debug=1  --mem pam_ipv6
Logging: /home/dan/repos/mariadb-server-10.3/mysql-test/mysql-test-run.pl  --mysqld=--loose-pam-debug=1 --mem pam_ipv6
vardir: /home/dan/repos/build-mariadb-server-10.3-debug/mysql-test/var
Checking leftover processes...
Removing old var directory...
Creating var directory '/home/dan/repos/build-mariadb-server-10.3-debug/mysql-test/var'...
 - symlinking 'var' to '/dev/shm/var_auto_LpLE'
Checking supported features...
MariaDB Version 10.3.5-MariaDB-debug
 - SSL connections supported
 - binaries are debug compiled
Collecting tests...
Installing system database...

==============================================================================

TEST                                      RESULT   TIME (ms) or COMMENT
--------------------------------------------------------------------------

worker[1] Using MTR_BUILD_THREAD 300, with reserved ports 16000..16019
plugins.pam_ipv6                         [ pass ]     30
--------------------------------------------------------------------------
The servers were restarted 0 times
Spent 0.030 of 4 seconds executing testcases

Completed: All 1 tests were successful.

$ grep PAM_RHOST  mysql-test/var/log/mysqld.1.err
PAM: pam_set_item(PAM_RHOST, ::1)
PAM: pam_set_item(PAM_RHOST, ::1)
<pre>

**Comment by svoj** [comment]
Hi @grooverdan,

Thanks for your contribution. Corresponding JIRA task: https://jira.mariadb.org/browse/MDEV-15547

This task was added to 10.3 backlog.

Thanks,
Sergey

**Comment by vuvova** [comment]
Generally, for the future, it'd be nice to see some explanation and justification for a pull request.

File: plugin/auth_pam/auth_pam.c [hunk 1]
```diff
@@ -220,7 +255,7 @@ maria_declare_plugin(pam)
   0x0100,
   NULL,
   vars,
-  "1.0",
+  "1.0.1",
   MariaDB_PLUGIN_MATURITY_STABLE
```
Comment (vuvova) [inline plugin/auth_pam/auth_pam.c]: that's very modest :) I'd go for at least 1.1

File: plugin/auth_pam/auth_pam.c [hunk 2]
```diff
@@ -19,6 +19,9 @@
 #include <mysql/plugin_auth.h>
 #include <stdio.h>
 #include <string.h>
+#include <sys/socket.h>
+#include <arpa/inet.h>
+#include <errno.h>
 #include <security/pam_appl.h>
```
Comment (vuvova) [inline plugin/auth_pam/auth_pam.c]: Are these three files guaranteed to exist when pam-devel headers are?

File: plugin/auth_pam/auth_pam.c [hunk 3]
```diff
@@ -150,6 +154,37 @@ static int pam_auth(MYSQL_PLUGIN_VIO *vio, MYSQL_SERVER_AUTH_INFO *info)
...
+      struct sockaddr_in6 *s6= (struct sockaddr_in6 *) &addr;
+      ip_str_ptr= inet_ntop(addr.ss_family,
+                            addr.ss_family==AF_INET ? (const void *) &s4->sin_addr : (const void *) &s6->sin6_addr,
+                            ip_str,
+                            sizeof(ip_str));
+      if (ip_str_ptr)
+      {
+        PAM_DEBUG((stderr, "PAM: pam_set_item(PAM_RHOST, %s)\n", ip_str_ptr));
+        DO( pam_set_item(pamh, PAM_RHOST, ip_str_ptr) );
+      }
```
Comment (vuvova) [inline plugin/auth_pam/auth_pam.c]: Why do you want to set `PAM_RHOST`? What's the use case?

Comment (grooverdan) [reply plugin/auth_pam/auth_pam.c]: PAM_RHOST provides information useful in diagnosis if there is an authentication failure. Like pam_tally2 (https://www.tecmint.com/use-pam_tally2-to-lock-and-unlock-ssh-failed-login-attempts/) can show you which host has the wrong authentication set.

Modules like pam_ihosts.so an restrict access based on IP.

With host information in the log, tools like fail2ban (https://github.com/fail2ban/fail2ban/blob/master/config/filter.d/pam-generic.conf) can block repeated failure attempts by IP.

File: plugin/auth_pam/auth_pam.c (hunk 2 above)
Comment (grooverdan) [reply plugin/auth_pam/auth_pam.c]: yes. tested on solaris, osx, freebsd and linux. There are also the ones needed for the syscalls performed, which come from POSIX standards

**Comment by grooverdan** [comment]
:snail:

**Pull request closed without merging**